packages/omgidl-serialization/src/MessageReader.ts

```diff
@@ -20,80 +26,142 @@ export type ArrayDeserializer = (
+      usesMemberHeader,
+    };
+
+    for (const field of complexDef.definitions) {
       if (field.isConstant === true) {
         continue;
       }
+      const definitionId = getDefinitionId(field);
+
+      let fieldLength = field.arrayLength;
+      if (readMemberHeader) {
+        const { id, objectSize: objectSizeBytes } = reader.emHeader();
+        const itemSize = typeToByteLength[field.type];
+        if (itemSize != undefined) {
+          fieldLength ??= Math.ceil(objectSizeBytes / itemSize);
+        }
+        // this can help spot misalignments in reading the data
+        if (definitionId != undefined && id !== definitionId) {
+          throw Error(
+            `CDR message deserializer error. Expected ${definitionId} but EMHEADER contained ${id} for field "${
+              field.name
+            }" in ${complexDef.name ?? "unknown"}`,
+          );
+        }
+      }
 
       if (field.isComplex === true) {
         // Complex type
-        const nestedDefinition = this.definitions.get(field.type);
-        if (nestedDefinition == undefined) {
+        const nestedComplexDef = this.definitions.get(field.type);
+        if (nestedComplexDef == undefined) {
           throw new Error(`Unrecognized complex type ${field.type}`);
         }
 
         if (field.isArray === true) {
           // For dynamic length arrays we need to read a uint32 prefix
-          const arrayLength = field.arrayLength ?? reader.sequenceLength();
+          const arrayLength = fieldLength ?? reader.sequenceLength();
           const array = [];
           for (let i = 0; i < arrayLength; i++) {
-            array.push(this.readComplexType(nestedDefinition, reader));
+            array.push(this.readComplexType(nestedComplexDef, reader, childOptions));
           }
           msg[field.name] = array;
         } else {
-          msg[field.name] = this.readComplexType(nestedDefinition, reader);
+          msg[field.name] = this.readComplexType(nestedComplexDef, reader, childOptions);
         }
       } else {
-        // Primitive type
         if (field.isArray === true) {
           const deser = typedArrayDeserializers.get(field.type);
           if (deser == undefined) {
             throw new Error(`Unrecognized primitive array type ${field.type}[]`);
           }
           // For dynamic length arrays we need to read a uint32 prefix
-          const arrayLength = field.arrayLength ?? reader.sequenceLength();
+          const arrayLength = fieldLength ?? reader.sequenceLength();
           msg[field.name] = deser(reader, arrayLength);
         } else {
           const deser = deserializers.get(field.type);
           if (deser == undefined) {
             throw new Error(`Unrecognized primitive type ${field.type}`);
           }
-          msg[field.name] = deser(reader);
+          msg[field.name] = deser(reader, fieldLength);
         }
       }
     }
     return msg;
   }
 }
 
+const typeToByteLength: Record<string, number> = {
```

Comment:
probably doesn't matter too much here but a function that has a switch might be faster - maybe not